UnitystationLauncher/ViewModels/SecurityPanelViewModel.cs

```diff
@@ -0,0 +1,538 @@
+            {
+                CopyFilesRecursively(goodRoot.ToString(), processingDirectory.ToString());
+            }
+            catch (Exception e)
+            {
+                errors.Invoke("e > " + e);
+                DeleteContentsOfDirectory(processingDirectory);
+                return;
+            }
+
+
+            var exeRename = processingDirectory.GetFiles()
+                .FirstOrDefault(x => x.Extension == ".exe" && x.Name != "UnityCrashHandler64.exe"); //TODO OS
+
+            if (exeRename == null || exeRename.Directory == null)
+            {
+                errors.Invoke("no Executable found ");
+                DeleteContentsOfDirectory(processingDirectory);
+                return;
+            }
+
+            exeRename.MoveTo(Path.Combine(exeRename.Directory.ToString(), dataPath.Name.Replace("_Data", "") + ".exe"));
+
+
+            DirectoryInfo goodBuilddirectory = directory.CreateSubdirectory("GoodBuild");
+            DeleteContentsOfDirectory(goodBuilddirectory);
+            CopyFilesRecursively(processingDirectory.ToString(), goodBuilddirectory.ToString());
+            DeleteContentsOfDirectory(processingDirectory);
+            //deleteContentsOfDirectory(Stagingdirectory); TODO
+        }
+        catch (Exception e)
+        {
+            errors.Invoke(" an Error happened > " + e);
+            DeleteContentsOfDirectory(processingDirectory);
+            //deleteContentsOfDirectory(Stagingdirectory); TODO
+        }
+    }
+
+    public bool ScanFolder(DirectoryInfo @unsafe, DirectoryInfo saveFiles, Action<string> info, Action<string> errors)
+    {
+        var goodFiles = saveFiles.GetFiles().Select(x => x.Name).ToList();
+
+        info.Invoke("Provided files " + string.Join(",", goodFiles));
+
+        CopyFilesRecursively(saveFiles.ToString(), @unsafe.ToString());
+
+        var files = @unsafe.GetFiles();
+
+
+        List<string> multiAssemblyReference = new List<string>();
+
+        foreach (var file in files)
+        {
+            if (goodFiles.Contains(file.Name)) continue;
+            multiAssemblyReference.Add(Path.GetFileNameWithoutExtension(file.Name));
+        }
+
+
+        foreach (var file in files)
+        {
+            if (goodFiles.Contains(file.Name)) continue;
+
+            info.Invoke("Scanning " + file.Name);
+
+            try
+            {
+                var listy = multiAssemblyReference.ToList();
+                listy.Remove(Path.GetFileNameWithoutExtension(file.Name));
+                if (_IAssemblyChecker.CheckAssembly(file, @unsafe, listy, info, errors) == false)
+                {
+                    errors.Invoke($"{file.Name} Failed scanning Cancelling");
+                    return false;
+                }
+            }
+            catch (Exception e)
+            {
+                errors.Invoke(" Failed scan due to error of " + e);
+                return false;
+            }
+        }
+
+
+        return true;
+    }
+
+
+    #region Utilities
```

Comment:
Lot of file operations happening in this region, preferably they should be in a service...

Reply:
this is moved to code scan service, shall I do the same thing there?